### PR TITLE
auto_source: optimize api calls and use head to determine if file exists

### PR DESCRIPTION
**Pull request by jnewb1**
auto_source was quite slow because it queried the api several times and also downloaded the file to determine if it existed

### CODE REVIEW

File: tools/lib/filereader.py [hunk 1]
```diff
@@ -8,6 +10,12 @@ def resolve_name(fn):
     return fn.replace("cd:/", DATA_ENDPOINT)
   return fn
 
+def file_exists(fn):
+  fn = resolve_name(fn)
+  if fn.startswith(("http://", "https://")):
+    return requests.head(fn).status_code == 200
+  return os.path.exists(fn)
```
Comment (sshane) [inline tools/lib/filereader.py:17]: Are you sure this works?

```
file_exists('cd:/297b4b460f361603/2023-07-25--03-01-46/0/rlog.bz2')
Out[3]: False
```

File: tools/lib/logreader.py [hunk 2]
```diff
@@ -173,10 +170,11 @@ def _parse_identifiers(self, identifier: str | List[str]):
...
 
     sr = SegmentRange(parsed)
+    route = Route(sr.route_name)
     mode = self.default_mode if sr.selector is None else ReadMode(sr.selector)
     source = self.default_source if source is None else source
 
-    return source(sr, mode)
+    return source(sr, route, mode)
 
   def __init__(self, identifier: str | List[str], default_mode=ReadMode.RLOG, default_source=auto_source, sort_by_time=False, only_union_types=False):
```
Comment (sshane) [inline tools/lib/logreader.py:179]: reset is run twice when using the log reader on one q/rlog one time, is this desired? It causes duplicate API queries

File: tools/lib/filereader.py (hunk 1 above)
Comment (jnewb1) [reply tools/lib/filereader.py]: needed allow_redirects, fixed

File: tools/lib/logreader.py (hunk 2 above)
Comment (jnewb1) [reply tools/lib/logreader.py:179]: reset in __iter__ isn't required anymore, fixed

**Pull request merged**